FlexMeasures/flexmeasures
# Add support for adding titles to asset page graphs

## Description

To allow for custom graph titles for sensors input by the user by modifying the `sensors_to_show` attribute in the attributes section.

Files changed:

- `flexmeasures/data/models/generic_assets.py`:
    - Modified `sensors_to_show` attribute to accept dictionaries with `title` and `sensor` keys or `sensors`.
    - Added logic to handle new format and populate `sensors_to_show` with `{'title': 'sensor_title', 'sensor': sensor_id}` or `{'title': 'sensor_title', 'sensors': List[sensor_ids]}`.
- `flexmeasures/data/schema_generic_assets.py`:
    - Refactored `GenericAssetSchema` to create seperate `SensorsToShowSchema`.
    - Added checks to validate the format of `title` and `sensor` or `sensors` fields in the dictionary.
- `flexmeasures/data/models/charts/belief_charts.py`:
    - Modified `chart_for_multiple_sensors` function to handle the new format.
    - Added support for custom titles in the graph rows.
- `flexmeasures/utils/coding_utils.py`:
    - Modified `flatten_unique` function to handle the new format and extract sensor IDs from dictionaries with `title` and `sensor` or `sensors` keys.

## Look & Feel

When `{'title': 'sensor_title', 'sensor': sensor_id}`  or  `{'title': 'sensor_title', 'sensors': list[sensor_ids]}` is added to `sensors_to_show`, this will set the custom title for the graph of that sensor.

## How to test

Manually test different JSON attributes in the edit asset sidebar. 
Verify that custom titles are correctly applied to the corresponding sensor graphs.

## Further Improvements

[] move the parsing functionality from flatten_unique to SensorsToShowSchema.
[] Add test case.
[] Add documentation and update the change-log for this feature.


## Related Items
 closes  #1119
---

- [x] I agree to contribute to the project under the Apache 2 License.
